crates/deltalake-core/src/operations/cast.rs

```diff
@@ -8,29 +8,27 @@ use std::sync::Arc;
 
 use crate::DeltaResult;
 
-fn cast_record_batch_columns(
-    batch: &RecordBatch,
+fn cast_struct(
+    struct_array: &StructArray,
     fields: &Fields,
     cast_options: &CastOptions,
 ) -> Result<Vec<Arc<(dyn Array)>>, arrow_schema::ArrowError> {
     fields
         .iter()
-        .map(|f| {
-            let col = batch.column_by_name(f.name()).unwrap();
-
+        .map(|field| {
+            let col = struct_array.column_by_name(field.name()).unwrap();
             if let (DataType::Struct(_), DataType::Struct(child_fields)) =
```

Comment:
There is one branch missing I guess and that's if the data type is List(Struct), so perhaps we need to add:

```rust
else if (DataType::List(_), DataType::List(to_inner_field)) { 
    match to_inner_field.data_type()  {
         Struct(child_fields) => .... do struct casting here,
         _ => ... continue normal flow
}
}

```

Reply:
I think this should be a follow up PR. Primary scope for this is to remove the panic / unsoundness that occurred with null-able structs. If a `List<Struct>` is provided and LHS does not match RHS then a proper error value is returned by `cast_with_options`.

Reply:
Alright let's have this in a separate PR :)